gems/sorbet/test/snapshot/update_one_bazel.sh

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+shopt -s dotglob
+
+root_dir="$PWD"
+
+# shellcheck disable=SC1090
+source "$root_dir/gems/sorbet/test/snapshot/logging.sh"
+
+# ----- Option parsing -----
+
+archive_path=$1
+update_path=$2
+test_name=$3
+
+test_dir="${root_dir}/gems/sorbet/test/snapshot/${test_name}"
+
+if [[ "${test_name}" =~ partial/* ]]; then
+  is_partial=1
+else
+  is_partial=
+fi
+
+info "├─ archive_path: $archive_path"
+info "├─ update_path: $update_path"
+info "├─ test_name: $test_name"
+info "├─ is_partial: ${is_partial:-}"
+info "├─ test_dir: $test_dir"
+
+(
+  cd "$test_dir"
+
+  # ----- Unpack the results archive -----
+  info "├─ Unpacking test archive"
+  tar -xvf actual.tar.gz
+
+
+  # ----- Update sorbet/ -----
+
+  diff_total() {
+    if ! diff -ur "expected/sorbet" "actual/sorbet"; then
+      attn "├─ updating expected/sorbet (total):"
+      rm -rf "$test_dir/expected/sorbet"
+      mkdir -p "$test_dir/expected"
+      cp -r "actual/sorbet" "$test_dir/expected"
+    fi
+  }
+
+  diff_partial() {
+    set +e
+    diff -ur "expected" "actual" | \
+      grep -vF "Only in actual" \
+      > "actual/partial-diff.log"
+    set -e
+
+    # File exists and is non-empty
+    if [ -s "actual/partial-diff.log" ]; then
+      attn "├─ updating expected/sorbet (partial):"
+
+      find "$test_dir/expected/sorbet" -print0 | while IFS= read -r -d '' expected_path; do
+        path_suffix="${expected_path#$test_dir/expected/sorbet}"
+        actual_path="actual/sorbet$path_suffix"
+
+        # Only ever update existing files, never grow this partial snapshot.
+        if [ -d "$expected_path" ]; then
+          if ! [ -d "$actual_path" ]; then
+            info "├─ removing dir: $expected_path"
+            rm -rfv "$expected_path"
+          fi
+        else
+          if [ -f "$actual_path" ]; then
+            info "├─ updating file: $expected_path"
+            cp -v "$actual_path" "$expected_path"
+          else
+            info "├─ removing file: $expected_path"
+            rm -fv "$expected_path"
+          fi
+        fi
+      done
+    fi
+  }
+
+  if [ -d expected/sorbet ]; then
+    if [ "$is_partial" = "" ]; then
+      info "├─ Performing total diff"
+      diff_total
+    else
+      info "├─ Performing partial diff"
+
+      # we know that there's an expected dir, as that is a pre-requesite for this
+      # test to be run by bazel
+      diff_partial
+    fi
+  else
+    attn "├─ missing expected/sorbet directory, skipping update"
```

Comment:
Can you say something here about what the workflow is for recording a new test? i.e., imagine that I just added a single `Gemfile` to a new folder `partial/foo`, and I want to populate the expected folder automatically. Right now, you can pass `--update --record` which will have the behavior that a partial test is treated like a total test for the sake of updating it only. After that, it's easy to go remove the files that aren't relevant to this test, and keep the ones that are.

iiuc, after this change, to replicate this workflow, someone actually has to create a new total test, run update, and then make it a partial test + remove things.

Reply:
There isn't a great workflow right now. I think the best path forward is to add a script that builds the `actual.tar.gz` artifact for a test, and then manually moves things into the `expected` directory. Attempting to do this from within bazel will be difficult.